File: src/my_py_pkg/my_py_pkg/pbvs_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from cv_bridge import CvBridge
from sensor_msgs.msg import Image
from std_msgs.msg import Float64MultiArray, Float64
import cv2
import numpy as np
from my_py_pkg.vision_utils import (  # Updated import with package prefix
    compute_homography,
    recover_from_homography,
    estimate_homography,
    rodriguez,
    pbvs_controller
)

class PBVSNode(Node):
    def __init__(self):
        super().__init__('pbvs_node')
        
        # Declare parameters with correct default path
        self.declare_parameter('reference_image', 
                             '/home/c/ros2_ws/src/my_cpp_pkg/image_reference/reference_image.jpg')
        self.declare_parameter('lambdav', 0.5)
        self.declare_parameter('lambdaw', 0.5)
        
        # Initialize CV bridge
        self.bridge = CvBridge()
        
        # Load reference image with error handling
        ref_path = self.get_parameter('reference_image').value
        self.ref_image = cv2.imread(ref_path, cv2.IMREAD_COLOR)
        if self.ref_image is None:
            self.get_logger().error(f"Failed to load reference image: {ref_path}")
            raise RuntimeError("Reference image not found")
            
        # Camera calibration matrix (should be parameterized)
        self.K = np.array([
            [554.382713, 0.0, 320.0],
            [0.0, 554.382713, 240.0],
            [0.0, 0.0, 1.0]
        ])
        
        # Initialize state variables
        self.current_image = None
        self.H = np.eye(3)
        self.R = np.eye(3)
        self.t = np.zeros((3, 1))
        self.n = np.zeros(3)
        self.d = 0.0
        self.counter = 0
        self.homography_solution = None
        
        # Setup QoS for image transport (ROS2-compliant)
        image_qos = rclpy.qos.QoSProfile(
            reliability=rclpy.qos.ReliabilityPolicy.BEST_EFFORT,
            history=rclpy.qos.HistoryPolicy.KEEP_LAST,
            depth=1
        )
        
        # Create subscribers
        self.image_sub = self.create_subscription(
            Image,
            '/iris/usb_cam/image_raw',  # Verify correct topic name
            self.image_callback,
            image_qos
        )
        
        # Create publishers with appropriate QoS
        self.homography_pub = self.create_publisher(Float64MultiArray, '/homography_numerical', 10)
        self.ck_t_ct_pub = self.create_publisher(Float64MultiArray, '/ck_t_ct', 10)
        self.n_pub = self.create_publisher(Float64MultiArray, '/n_plane_vector', 10)
        self.Uv_pub = self.create_publisher(Float64MultiArray, '/Uv', 10)
        self.Uw_pub = self.create_publisher(Float64MultiArray, '/Uw', 10)
        self.d_pub = self.create_publisher(Float64, '/d_value', 10)
        
        # Processing timer (10Hz)
        self.process_timer = self.create_timer(0.1, self.process_callback)

    def image_callback(self, msg):
        try:
            self.current_image = self.bridge.imgmsg_to_cv2(msg, 'bgr8')
        except Exception as e:
            self.get_logger().error(f'Image conversion failed: {str(e)}')

    def process_callback(self):
        if self.current_image is None:
            return
            
        try:
            # Estimate homography
            self.H = estimate_homography(
                self.ref_image, 
                self.current_image, 
                self.K, 
                self.counter
            )
            
            if self.H is not None:
                # Recover pose from homography
                (self.R, self.t, self.n, self.d, 
                 self.homography_solution) = recover_from_homography(
                    self.H, 
                    self.K, 
                    self.counter, 
                    self.homography_solution
                )
                
                # Compute control vectors
                u = rodriguez(self.R)
                lambdav = self.get_parameter('lambdav').value
                lambdaw = self.get_parameter('lambdaw').value
                Uv, Uw = pbvs_controller(self.R, self.t, u, lambdav, lambdaw)
                
                # Publish results
                self.publish_matrix(self.homography_pub, self.H)
                self.publish_matrix(self.ck_t_ct_pub, self.t)
                self.publish_matrix(self.n_pub, self.n)
                self.d_pub.publish(Float64(data=self.d))
                self.publish_matrix(self.Uv_pub, Uv)
                self.publish_matrix(self.Uw_pub, Uw)
                
            self.counter += 1
            
        except Exception as e:
            self.get_logger().error(f'Processing error: {str(e)}', throttle_duration_sec=1.0)

    def publish_matrix(self, publisher, matrix):
        """Helper function to publish numpy matrices as Float64MultiArray"""
        msg = Float64MultiArray()
        msg.data = matrix.flatten().tolist()
        publisher.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    
    try:
        node = PBVSNode()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    except Exception as e:
        node.get_logger().error(f'Node initialization failed: {str(e)}')
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/my_cpp_pkg/src/pbvs_node.cpp
#include "my_cpp_pkg/vision_utils.hpp"
#include <Eigen/Dense>

#include <rclcpp/rclcpp.hpp>
#include "geometry_msgs/msg/twist.hpp"
#include "image_transport/image_transport.hpp"
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <iostream>
#include <stdlib.h>

#include "std_msgs/msg/float64_multi_array.hpp"
#include "std_msgs/msg/float64.hpp"

using namespace std::chrono_literals;

cv::Mat refImage, currentImage;
double t0,t;

static const std::string REF_IMAGE_NAME = "/home/c/ros2_ws/src/my_cpp_pkg/image_reference/reference_image.jpg";

class PBVSNode : public rclcpp::Node {
public:
    PBVSNode() : Node("pbvs_node") {
        refImage = cv::imread(REF_IMAGE_NAME, cv::IMREAD_COLOR);
        if(refImage.empty()) {
            RCLCPP_ERROR(this->get_logger(), "Could not read the image: %s", REF_IMAGE_NAME.c_str());
            rclcpp::shutdown();
            exit(1);
        } else {
            RCLCPP_INFO(this->get_logger(), "Reference image is ready: %s", REF_IMAGE_NAME.c_str());
        }

        topic_name_ = this->declare_parameter<std::string>("topic_name", "/my_image");
        lambdav_ = this->declare_parameter<double>("lambdav", 1.0);
        lambdaw_ = this->declare_parameter<double>("lambdaw", 1.0);

        homography_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/homograpy_numerical", 10);
        ck_t_ct_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/ck_t_ct", 10);
        n_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/n_plane_vector", 10);
        Uv_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/Uv", 10);
        Uw_pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/Uw", 10);
        d_pub_ = this->create_publisher<std_msgs::msg::Float64>("/d_value", 10);

        image_transport::ImageTransport it(this);
        image_subscription_ = it.subscribe(topic_name_, 1, std::bind(&PBVSNode::imageCallback, this));

        timer_ = this->create_wall_timer(100ms, std::bind(&PBVSNode::controlLoop, this));
    }

private:
    void imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr & msg) {
        cv_bridge::CvImagePtr cvPtr;
        try {
            cvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        } catch (cv_bridge::Exception &e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }
        cvPtr->image.copyTo(currentImage);
    }

    void controlLoop() {
        if(currentImage.empty()) {
            RCLCPP_INFO(this->get_logger(), "Waiting for current image...");
            return;
        }

        cv::imwrite("/tmp/current_image.jpg", currentImage);

        Matrix_t K(3,3);
        Matrix_t H(3,3);
        Matrix_t R(3,3);
        Vector3D_t t;
        Vector3D_t n;
        Vector3D_t u;
        Vector3D_t Uv;
        Vector3D_t Uw;
        FPTYPE d;
        homographySolution homSolution;
        static int counter = 0;

        K << 554.382713,    0.0     , 320.0,
             0.0         , 554.382713, 240.0,
             0.0         ,     0.0     , 1.0;

        EstimateHomography(refImage, currentImage, K, H, counter );
        RecoverFromHomography(H,R,t,n,d, counter, homSolution);
        Rodriguez(R,u);
        PBVSController(R,t,u,Uv,Uw, lambdav_, lambdaw_);

        std_msgs::msg::Float64MultiArray H_msg, ck_t_ct_msg, n_msg, Uv_msg, Uw_msg;
        std_msgs::msg::Float64 d_msg;

        H_msg.data = std::vector<double>(H.data(), H.data() + H.size());
        ck_t_ct_msg.data = std::vector<double>(t.data(), t.data() + t.size());
        n_msg.data = std::vector<double>(n.data(), n.data() + n.size());
        Uv_msg.data = std::vector<double>(Uv.data(), Uv.data() + Uv.size());
        Uw_msg.data = std::vector<double>(Uw.data(), Uw.data() + Uw.size());

        d_msg.data = d;

        homography_pub_->publish(H_msg);
        ck_t_ct_pub_->publish(ck_t_ct_msg);
        n_pub_->publish(n_msg);
        d_pub_->publish(d_msg);
        Uv_pub_->publish(Uv_msg);
        Uw_pub_->publish(Uw_msg);

        counter++;
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr homography_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr ck_t_ct_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr n_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr Uv_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr Uw_pub_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr d_pub_;
    image_transport::Subscriber image_subscription_;
    std::string topic_name_;
    double lambdav_;
    double lambdaw_;
};


int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PBVSNode>());
    rclcpp::shutdown();
    return 0;
}
